### DISCUSSION TITLE
Some questions about Sensitivity evaluation via Casadi in Matlab

### DISCUSSION THREAD

**Discussion by ookuraresona** [casadi/casadi#3271] [Q&A]
Dear all,
  Sorry for your time. I am doing some experiments on distributed optimizaiton problem. And currently i am stuck by the sensitivity evaluation for nonconvex subproblems. Actually, i have the statement "Optimal solution found", which means i get the local minimizer of the problem. Then, i use the Matlab function "gradient" and "hessian" to have the function of gradient and hessian of the Lagrangian function and test the value of each function at the local minimizer. But what i get is that the value of  the gradient at local minimizer isn't equal to zeros and the reduced hessian matrix isn't positive definite (violate the SOSC).
  I am wondering that if i misunderstand the KKT condition or use the function in a wrong way. Could you plz help me how to evaluate the dual feasibility and sensitivity at the local minimizer? Here are some statements below.

"Number of nonzeros in equality constraint Jacobian...:      567
Number of nonzeros in inequality constraint Jacobian.:      240
Number of nonzeros in Lagrangian Hessian.............:      607

Total number of variables............................:      157
                     variables with only lower bounds:       33
                variables with lower and upper bounds:      104
                     variables with only upper bounds:        0
Total number of equality constraints.................:      122
Total number of inequality constraints...............:       35
        inequality constraints with only lower bounds:       15
   inequality constraints with lower and upper bounds:       20
        inequality constraints with only upper bounds:        0

                                   (scaled)                 (unscaled)
Objective...............:  3.1924468481628019e+003   1.0189331015845902e+006
Dual infeasibility......:  6.5964649800970587e-008   2.1053933992769379e-005
Constraint violation....:  2.3496970889315018e-009   2.3496970889315018e-009
Complementarity.........:  1.5947592682477112e-009   5.0899923624775280e-007
Overall NLP error.......:  2.3496970889315018e-009   2.1053933992769379e-005

Number of objective function evaluations             = 637
Number of objective gradient evaluations             = 224
Number of equality constraint evaluations            = 637
Number of inequality constraint evaluations          = 637
Number of equality constraint Jacobian evaluations   = 225
Number of inequality constraint Jacobian evaluations = 225
Number of Lagrangian Hessian evaluations             = 223
Total CPU secs in IPOPT (w/o function evaluations)   =      0.431
Total CPU secs in NLP function evaluations           =      0.587

EXIT: Optimal Solution Found.
      solver  :   t_proc      (avg)   t_wall      (avg)    n_eval
       nlp_f  |  65.00ms (102.04us)  65.00ms (102.04us)       637
       nlp_g  |  79.00ms (124.02us)  79.02ms (124.05us)       637
    nlp_grad  |        0 (       0)        0 (       0)         1
  nlp_grad_f  |  57.00ms (253.33us)  56.99ms (253.29us)       225
  nlp_hess_l  | 299.00ms (  1.35ms) 299.12ms (  1.35ms)       222
   nlp_jac_g  |  81.00ms (358.41us)  80.99ms (358.37us)       226
       total  |   1.02 s (  1.02 s)   1.02 s (  1.02 s)         1
"
"     kkappCas = SX.sym('kapp',lengthLocCons{i},1);
       gradiLag     = gradient(sProb.obj_block{i}  + kkappCas'* ...
            sProb.locfunConstraints{i},sProb.xxCas{i});
       HHiCas    = hessian(sProb.obj_block{i} + kkappCas'* ...
                sProb.locfunConstraints{i}, sProb.xxCas{i});
"
Thanks a lot!

**Comment by brunoml5** [comment]
Hi!

I'm a beginner in this topic but I saw some stuff in this [blog post](https://web.casadi.org/blog/nlp_sens/) and the [article](https://www.sciencedirect.com/science/article/pii/S2405896318327137) from the developers.

You are using IPOPT which is a solver based on the barrier method. Because of that, it does not properly find the active-set, i.e. the active contraints in the problem. It only converges up to a small tolerance. 

IPOPT is robust but is not accurate for the dual variables. So, one interesting option is to solve the problem with IPOPT and then use the solution as a warm start for an active-set solver which will be more accurate for the sensitivity variables. In the [article](https://www.sciencedirect.com/science/article/pii/S2405896318327137), the developers explain how the QR-SQP works.

Try this and tell us if it returns a better result.

I hope it helps!

**Reply by ookuraresona** [reply]
Hi!

Thanks a lot for your suggestion. The article looks like what i am looking for and i'll go through it quickly. Actually, in IPOPT the "dual infeasibilty" state the gradient of the Lagrangian function at current point. Maybe if i can get the function of evaluating "dual infeasibility" in IPOPT and then i can solve the problem:)
